### PR TITLE
Add public API to vote for custom events

### PR DESCRIPTION
**Pull request by nbaksalyar**
This is required to achieve consensus on events in Vaults.

Closes https://github.com/maidsafe/safe_vault/issues/868

### CODE REVIEW

File: src/states/elder/mod.rs [hunk 1]
```diff
@@ -373,7 +373,7 @@ impl Elder {
                 }
 
                 // Drop: condition may have changed.
-                AccumulatingEvent::OurMerge => false,
+                AccumulatingEvent::OurMerge | AccumulatingEvent::Custom(_) => false,
```
Comment (madadam) [inline src/states/elder/mod.rs]: I'm not sure about this, but we might want to carry the custom events through prefix changes.

File: src/chain/network_event.rs [hunk 2]
```diff
@@ -80,6 +80,9 @@ pub enum AccumulatingEvent {
 
     // Prune the gossip graph.
     ParsecPrune,
+
+    // Opaque user-defined event.
+    Custom(Vec<u8>),
```
Comment (nbaksalyar) [inline src/chain/network_event.rs]: A better name was suggested for this event, so I'll be making this change:
```suggestion
    User(Vec<u8>),
```

File: src/states/adult.rs [hunk 3]
```diff
@@ -504,6 +504,14 @@ impl Approved for Adult {
         Ok(())
     }
 
+    fn handle_custom_event(
+        &mut self,
+        _payload: Vec<u8>,
+        _outbox: &mut dyn EventBox,
+    ) -> Result<(), RoutingError> {
+        Ok(())
```
Comment (madadam) [inline src/states/adult.rs]: Shouldn't we raise these also as adult?

File: src/states/elder/mod.rs (hunk 1 above)
Comment (nbaksalyar) [reply src/states/elder/mod.rs]: >we might want to carry the custom events through prefix changes.

Wasn't sure about this either. Probably yes, it's better to leave the filtering to a user: so e.g. in the Vaults case we can decide if these events are still relevant to us if we know the prefix has changed.

Comment (madadam) [reply src/states/elder/mod.rs]: Should work. Assuming vaults get notified about prefix changes.

File: src/states/elder/mod.rs [hunk 4]
```diff
@@ -1349,6 +1355,15 @@ impl Approved for Elder {
...
     }
+
+    fn handle_user_event(
+        &mut self,
+        payload: Vec<u8>,
+        outbox: &mut dyn EventBox,
+    ) -> Result<(), RoutingError> {
+        self.send_event(Event::Consensus(payload), outbox);
+        Ok(())
+    }
```
Comment (fizyk20) [inline src/states/elder/mod.rs]: The implementations are the same in both `Adult` and `Elder` - wouldn't it be better to make this the default implementation in the `Approved` trait?

Comment (nbaksalyar) [reply src/states/elder/mod.rs]: Agreed, good suggestion :+1:

**Comment by nbaksalyar** [comment]
Thanks for the review! Rebased & force-pushed now :slightly_smiling_face:

**Pull request merged**